pkg/distributor/otel.go

```diff
@@ -335,28 +342,61 @@ func httpRetryableToOTLPRetryable(httpStatusCode int) int {
+// writeErrorToHTTPResponseBody converts the given error into a gRPC status and marshals it into a byte slice, in order to be written to the response body.
+// See doc https://opentelemetry.io/docs/specs/otlp/#failures-1.
+func writeErrorToHTTPResponseBody(r *http.Request, w http.ResponseWriter, httpCode int, grpcCode codes.Code, msg string, logger log.Logger) {
 	validUTF8Msg := validUTF8Message(msg)
-	w.Header().Set("Content-Type", "application/octet-stream")
-	w.Header().Set("X-Content-Type-Options", "nosniff")
-	if server.IsHandledByHttpgrpcServer(reqCtx) {
+	if server.IsHandledByHttpgrpcServer(r.Context()) {
 		w.Header().Set(server.ErrorMessageHeaderKey, validUTF8Msg) // If httpgrpc Server wants to convert this HTTP response into error, use this error message, instead of using response body.
 	}
-	w.WriteHeader(httpCode)
 
-	respBytes, err := proto.Marshal(status.New(grpcCode, validUTF8Msg).Proto())
-	if err != nil {
-		level.Error(logger).Log("msg", "otlp response marshal failed", "err", err)
-		writeResponseFailedBody, _ := proto.Marshal(status.New(codes.Internal, "failed to marshal OTLP response").Proto())
-		_, _ = w.Write(writeResponseFailedBody)
-		return
+	st := status.New(grpcCode, validUTF8Msg).Proto()
+	writeOTLPResponse(r, w, httpCode, st, logger)
+}
+
+func writeOTLPResponse(r *http.Request, w http.ResponseWriter, httpCode int, payload proto.Message, logger log.Logger) {
+	contentType := r.Header.Get("Content-Type")
+	switch contentType {
+	case jsonContentType, pbContentType:
+	default:
+		// Default to protobuf encoding.
+		contentType = pbContentType
 	}
 
-	_, err = w.Write(respBytes)
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+
+	var body []byte
+	var err error
+	if contentType == jsonContentType {
+		body, err = json.Marshal(payload)
+		if err != nil {
+			httpCode = http.StatusInternalServerError
+			level.Error(logger).Log("msg", "failed to marshal payload to JSON", "err", err, "payload", payload)
+			body, err = json.Marshal(status.New(codes.Internal, "failed to marshal OTLP response").Proto())
+			err = errors.Wrapf(err, "marshalling %T to JSON", payload)
+		}
+	} else {
+		body, err = proto.Marshal(payload)
+		if err != nil {
+			httpCode = http.StatusInternalServerError
+			level.Error(logger).Log("msg", "failed to marshal payload to protobuf", "err", err, "payload", payload)
+			body, err = proto.Marshal(status.New(codes.Internal, "failed to marshal OTLP response").Proto())
+			err = errors.Wrapf(err, "marshalling %T to protobuf", payload)
+		}
+	}
 	if err != nil {
-		level.Error(logger).Log("msg", "write error to otlp response failed", "err", err)
+		level.Error(logger).Log("msg", "OTLP response marshal failed, responding without payload", "err", err, "content_type", contentType)
+		contentType = "text/plain"
+		body = nil
+	}
+
+	w.Header().Set("Content-Type", contentType)
```

Comment:
Do we also need to set `Content-Length`?

Reply:
The OTLP spec doesn't say anything about Content-Length. As such, let's keep it out of the scope of this PR?

Reply:
It's good practice to set it where possible, it allows clients to size buffers appropriately etc.

Reply:
I'll do it in a follow-up, to separate changes.